### DISCUSSION TITLE
Add support to link on official documentation to TextReporter

### DISCUSSION THREAD

**Discussion by llaville** [PHP-CS-Fixer/PHP-CS-Fixer#7996] [Feature ideas]
Hello,

If you are a senior developer/user of PHP-CS-Fixer, it's probably easy to know what each fixer is supposed to do, but for other audience it's probably not the case !

This is the reason why I've worked today on a fix that could be include to official TextReporter.
An alternative will be to allow to register our own reporters (see https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/discussions/7995).

Here is what we could have as an improved version of TextReporter.

Standard way is same 

```
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

   1) src/SarifLog.php

Found 1 of 1 files that can be fixed in 0.034 seconds, 16.000 MB memory used
```
 
With verbose level 1 is same 

```
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

   1) src/SarifLog.php (blank_line_after_opening_tag, ternary_operator_spaces, blank_lines_before_namespace, blank_line_between_import_groups)

Found 1 of 1 files that can be fixed in 0.028 seconds, 16.000 MB memory used
```

Add extra info (link to official doc) with verbose level 2 : 

```
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

   1) src/SarifLog.php
      blank_line_after_opening_tag
      : https://cs.symfony.com/doc/rules/php_tag/blank_line_after_opening_tag.html
      ternary_operator_spaces
      : https://cs.symfony.com/doc/rules/operator/ternary_operator_spaces.html
      blank_lines_before_namespace
      : https://cs.symfony.com/doc/rules/namespace_notation/blank_lines_before_namespace.html
      blank_line_between_import_groups
      : https://cs.symfony.com/doc/rules/whitespace/blank_line_between_import_groups.html

Found 1 of 1 files that can be fixed in 0.028 seconds, 16.000 MB memory used
```

WDYT about this feature idea ?

PS: I've already a fix working

**Comment by Wirone** [comment]
Interesting idea, but it would take up much space in the output. It would be much better if each rule name was clickable (ctrl+click, just like when you use `editorLink` in PHPStan and can open a reported error in the IDE). I know it may be less discoverable, but at the same time it won't be so... chaotic 😅? What do you think?

**Reply by llaville** [reply]
It's already the case. I've follow same principle as implemented in `composer show` command (see https://github.com/composer/composer/pull/10430).
You can give it a try if you want to see it in action : see my fork repo https://github.com/llaville/PHP-CS-Fixer/tree/features/add_links_to_doc

**Reply by Wirone** [reply]
I am AFK so I can't check, I just read your proposal and there's nothing about clickable names, and the last output (verbosity l.2) shows URLs directly, which is too verbose IMHO. I'll try to check your branch soon.

**Reply by Wirone** [reply]
Just as I thought looking at your comment, the links are displayed _and_ are clickable:

<img width="792" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/0dba1f1f-3ef5-43cf-987f-b65fa02317bc">

What I said, is something like this:

<img width="956" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/ab837d23-0f89-4ec0-8a39-f32f15bab354">

which means exactly the same output as currently, but with clickable rules 🙂. I did it with `sprintf('<href=%s;fg=yellow>%s</>', OutputFormatter::escape($url), $appliedFixer)`. That is a great addition IMHO, please rework it to not change the items' alignment in the output, but only to add `href` support, and create PR with it.

**Reply by Wirone** [reply]
Ah, and rules should be clickable when `--verbose` / `-v` is present and applied rules are displayed, no need to enable it for `-vv` only 😊.

**Reply by llaville** [reply]
In this last case, there are many unit tests to change because it will introduced regression ! Are you aware of this ?

**Reply by llaville** [reply]
Ok all fixes related to this new feature was applied: PR #8003 is opened

**Reply by Wirone** [reply]
> In this last case, there are many unit tests to change because it will introduced regression ! Are you aware of this ?

Tests for the output should use non-decorated mode, so the link is stripped from them and this change should not affect previous tests. There should be a dedicated test to ensure href is present in decorated mode. Is there really regression?

**Reply by llaville** [reply]
Probably, or I missed something : See PR results https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/9041203007?pr=8003

**Reply by llaville** [reply]
Yes, at least I missed to fix RunnerTest ... on way !

**Reply by llaville** [reply]
Done with commit https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8003/commits/48b82ca73eeaa74b84f1429895b7c895e54ce810

**Reply by llaville** [reply]
@Wirone All is OK now : ready for code review on PR #8003

**Reply by llaville** [reply]
@Wirone New version with my last changes (see comment https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8003#issuecomment-2105977311) is available: PR ready for review !

**Reply by Wirone** [reply]
No need to ping me, I have notifications enabled for the project so I know about every update 🙂.
